allow the user to set the RNG seed in distort_random()

dfcastellanos opened the pull request:
The function distort_random induces a random distortion of the grid, using the default initialization of the std::mt19937 RNG used. Consequently, this function does not allow for different random distortions of the same mesh, which might be necessary for e.g. statistical averaging of the results over many different distortions. 

To solve this issue, a seed parameter is added to distort_random(). To not break the existing interface, the seed parameter has a default value. Moreover, the default seed is 5489u, which is the default value for std::mt19937 (see http://www.cplusplus.com/reference/random/mt19937/). Thus, when no seed is provided, the distortion will be numerically the same as in previous versions.

masterleinad commented:
To fix the indentation it should be sufficient to just run `make inden` or `./contrinb/utilities/indent`.

masterleinad commented:
Let us know if you need more help with that.

dfcastellanos commented:
> To fix the indentation it should be sufficient to just run `make inden` or `./contrinb/utilities/indent`.

Thank you, I was not aware of that utility.

bangerth commented on source/grid/grid_tools.cc:
```diff
@@ -1201,14 +1202,8 @@ namespace GridTools
-    // get the same mesh. this would not be the case if you used
-    // the rand() function, which carries around some internal state
-    // we could use std::mt19937 but doing so results in compiler-dependent
-    // output.
```
Would you mind re-instating this comment? I think it describes relevant behavior.

dfcastellanos replied:
The reason why I removed the comment is that it says that the result is reproducible when the function is applied over the same mesh since it uses mt19937 instead of rand(). However, that behavior is not given by mt19937 instead of rand(), but because it uses mt19937 with its default initialization, which happens to be deterministic (i.e., has a default seed). 

With the new seed parameter, that piece of comment is outdated since obtaining or not the same result is defined by the value of the seed, over which the user has control in the function call.

If you think this clarification should be added to the code, let me know.

bangerth replied:
Right. I think what I want to say is that "with the same seed, the function produces the same reproducible output", which is not true when using `rand()`. But let's not worry about it. I can always re-introduce the comment later on in slightly modified form.

bangerth commented on source/grid/grid_tools.inst.in:
```diff
@@ -254,7 +254,8 @@ for (deal_II_dimension : DIMENSIONS; deal_II_space_dimension : SPACE_DIMENSIONS)
         Triangulation<deal_II_dimension, deal_II_space_dimension> &,
-        const bool);
+        const bool,
+        unsigned int);
```
```suggestion
        const unsigned int);
```

masterleinad commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -628,12 +628,17 @@ namespace GridTools
                  Triangulation<dim, spacedim> &triangulation,
-                 const bool                    keep_boundary = true);
+                 const bool                    keep_boundary = true,
+                 const unsigned int            seed          = 5489u);
```
```suggestion
                 const unsigned int            seed          = boost::random::mt19937::default_seed);
```

dfcastellanos commented:
There's a new problem in the indent test, but it talks about user without firstname and lastname. I get the same when I run dealii/contrib/utilities/indent. Username and email have already been set. What's the problem?

masterleinad commented:
> There's a new problem in the indent test, but it talks about user without firstname and lastname. I get the same when I run dealii/contrib/utilities/indent. Username and email have already been set. What's the problem?

It seems that in one of your commits the author's name is not in the required format. Can you just squash all your commits and make sure that the information shown by `git log` for that one commit looks good?

dfcastellanos commented:
> > There's a new problem in the indent test, but it talks about user without firstname and lastname. I get the same when I run dealii/contrib/utilities/indent. Username and email have already been set. What's the problem?
> 
> It seems that in one of your commits the author's name is not in the required format. Can you just squash all your commits and make sure that the information shown by `git log` for that one commit looks good?

I squashed them and added authorship and email, and git log shows it. Still the indent test complains about the same issue.

jppelteret commented:
Hi @dfcastellanos 

Firstly, thanks a lot for your contribution!

> ... added authorship and email

Thanks, it looks like your name is now being correctly attributed to your commits, but

> I squashed them and...

unfortunately something's not quite right with how you did this step. If you look at the commits tab for your PR in GitHub, [here](https://github.com/dealii/dealii/pull/11003/commits), then you'll see that there are actually 11 commits on this branch. Squashing them all would reduce the number of commits to 1 (ideally with only the first commit message remaining). Given that you've now updated your git authorship, the remaining commit *should* have your updated details. Do you need a hand with this step?

dfcastellanos commented:
> Hi @dfcastellanos
> 
> Firstly, thanks a lot for your contribution!
> 
> > ... added authorship and email
> 
> Thanks, it looks like your name is now being correctly attributed to your commits, but
> 
> > I squashed them and...
> 
> unfortunately something's not quite right with how you did this step. If you look at the commits tab for your PR in GitHub, [here](https://github.com/dealii/dealii/pull/11003/commits), then you'll see that there are actually 11 commits on this branch. Squashing them all would reduce the number of commits to 1 (ideally with only the first commit message remaining). Given that you've now updated your git authorship, the remaining commit _should_ have your updated details. Do you need a hand with this step?

It worked now, thanks.

masterleinad commented:
> It worked now, thanks.

No, it didn't. You included changes from a bunch of other pull requests so that you are now changing roughly 1000 LOC. Please rebase on top of the upstream `master` branch.

dfcastellanos commented:
> > It worked now, thanks.
> 
> No, it didn't. You included changes from a bunch of other pull requests so that you are now changing roughly 1000 LOC. Please rebase on top of the upstream `master` branch.

Should I do 'git pull --rebase'? Or which would be the most appropriate command to solve this specific problem?

masterleinad commented:
> Should I do 'git pull --rebase'? Or which would be the most appropriate command to solve this specific problem?

I would do
```
git fetch upstream
git rebase upstream/master
```
assuming that you called the remote `git@github.com:dealii/dealii` (or `https://github.com/dealii/dealii`) `upstream`.

dfcastellanos commented:
> > Should I do 'git pull --rebase'? Or which would be the most appropriate command to solve this specific problem?
> 
> I would do
> 
> ```
> git fetch upstream
> git rebase upstream/master
> ```
> 
> assuming that you called the remote `git@github.com:dealii/dealii` (or `https://github.com/dealii/dealii`) `upstream`.

I did that and I get this merging problem:
```
First, rewinding head to replay your work on top of it...
Applying: allow the user to set the RNG seed in distort_random() - squashed previous commits
Using index info to reconstruct a base tree...
M	include/deal.II/dofs/dof_accessor.templates.h
M	include/deal.II/fe/mapping_q_generic.h
M	include/deal.II/grid/cell_id.h
M	include/deal.II/grid/grid_tools.h
M	include/deal.II/matrix_free/cuda_hanging_nodes_internal.h
M	include/deal.II/matrix_free/matrix_free.templates.h
M	include/deal.II/multigrid/mg_constrained_dofs.h
M	include/deal.II/numerics/matrix_creator.templates.h
M	include/deal.II/numerics/vector_tools_boundary.templates.h
M	include/deal.II/numerics/vector_tools_constraints.templates.h
M	include/deal.II/numerics/vector_tools_project.templates.h
M	source/base/polynomial.cc
M	source/dofs/dof_handler.cc
M	source/dofs/dof_handler_policy.cc
M	source/dofs/dof_tools.cc
M	source/dofs/dof_tools_constraints.cc
M	source/dofs/dof_tools_sparsity.cc
M	source/fe/mapping_q_generic.cc
M	source/grid/CMakeLists.txt
M	source/grid/cell_id.cc
M	source/grid/grid_tools.cc
M	source/grid/grid_tools.inst.in
A	source/grid/grid_tools_2.cc
M	source/grid/grid_tools_dof_handlers.cc
M	source/grid/tria_accessor.cc
M	source/multigrid/mg_tools.cc
M	source/particles/particle_handler.cc
Falling back to patching base and 3-way merge...
Auto-merging source/grid/grid_tools.inst.in
Auto-merging source/grid/grid_tools.cc
CONFLICT (rename/rename): Rename "source/grid/grid_tools_2.cc"->"source/fe/mapping_fe_field_inst2.cc" in branch "HEAD" rename "source/grid/grid_tools_2.cc"->"tests/grid/cell_id_07.cc" in "allow the user to set the RNG seed in distort_random() - squashed previous commits"
Auto-merging source/fe/mapping_fe_field_inst2.cc and tests/grid/cell_id_07.cc, both renamed from source/grid/grid_tools_2.cc
Auto-merging tests/grid/cell_id_07.cc
Auto-merging include/deal.II/grid/grid_tools.h
error: Failed to merge in the changes.
Patch failed at 0001 allow the user to set the RNG seed in distort_random() - squashed previous commits
hint: Use 'git am --show-current-patch' to see the failed patch
Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
```

masterleinad commented:
Hmmm... Yes, that looks a little more difficult. If you don't mind, I can fix the merge conflict and push to your branch.

dfcastellanos commented:
> Hmmm... Yes, that looks a little more difficult. If you don't mind, I can fix the merge conflict and push to your branch.

Sure, do it. But maybe it is easier if I just create a new branch and make a new pull request the right way?

masterleinad commented:
Here you go!

masterleinad commented:
/rebuild
